### ISSUE TITLE
wf2 - map latest answer to `sex` from encounter to TEI

### ISSUE THREAD

**Issue by aleksa-krolls** [OpenFn/openfn-lime-pilot#38]
## Background, context, and business value
Sometimes `sex` is specified on an OMRS form/encounter in a later consultation. If answered, we will need to map this answer for sex from the encounter to the trackedEntityInstance record. 

## The specific request, in as few words as possible
If either the `mhpss baseline` form (encounter id: `6a3e1e0e-dd13-3465-b8f5-ee2d42691fe5`)  or `mhgap baseline` form (encounter id: `82db23a1-4eb1-3f3c-bb65-b7ebfe95b19b`) have `sex` specified, then map this value to the dhis2 `TEI` record. 
- Check the `mhpss baseline` encounter first --> and map answer if the `sex` concept is found.
- If not found, then check the `mhgap baseline` encounter --> and map answer if the `sex` concept is found.
- If an answer for `sex` is found in neither encounter, then no action required. 

See example code snippet below. If either encounter contains an `obs[x]` where `concept.uuid = 'ec42d68d-3e23-43de-b8c5-a03bb538e7c7'`, then map the `obs[x].uuid` answer using the [optionsMap](https://gist.github.com/aleksa-krolls/2f309595110239cfcc8251017cf11592) --> to dhis2 `attribute: qptKDiv9uPl`

```
"obs": [
 {
                    "uuid": "87d4e635-fdf4-447a-9a79-1bf89dde9344",
                    "display": "Sexe: Unknown",
                    "concept": {
                        "uuid": "ec42d68d-3e23-43de-b8c5-a03bb538e7c7",
                        "display": "Sexe",
```

## Workflow
wf2 --> I think you'll need to update step `6-create-events` to add a step to update the parent `TEI` again if an answer for `sex` is found in either of the 2 encounters. 

Your mapping will look something like this... 
https://github.com/OpenFn/openfn-lime-pilot/blob/90efe525dadd77fa0af13621eba92ca576e03e07/workflows/wf2/2-upsert-teis.js#L64-L65

## Output
`qptKDiv9uPl` is the only attribute you'll need to update for this TEI, so your final payload will probably  look like this...
```
program: 'w9MSPn5oSqp',
orgUnit: 'OPjuJMZFLop',
trackedEntityType: 'cHlzCA2MuEF',
 attributes: [
          {
            attribute: 'qptKDiv9uPl', //gender
            value: optionMap[encounter.latestAnswerSex],
          },
          {
            attribute: 'AYbfTPYMNJH', //OpenMRS Patient UID to use to upsert TEI
            value: patient.uuid,
          },
```

## Testing guidance
Try running wf2 with this cursor to get patient `{"lastRunDateTime": "2024-09-30T07:00:00.000Z"}` to test this with. 

## Toggl 
`MSF 2024`

**Comment by aleksa-krolls** [comment]
@mtuchi this is a tricky one... I suggest we pair on it later today

**Comment by aleksa-krolls** [comment]
hey @mtuchi for this one...
1. [Gist](https://gist.githubusercontent.com/aleksa-krolls/2f309595110239cfcc8251017cf11592/raw/79bb85933038f305f19b77ace8d807e8e1751ac6/msf_wf2_optionsMap.json) updated and new link added to staging
2. See "sex" diagram in [LucidChart](https://lucid.app/lucidchart/38ae80f0-0650-4fea-9a11-dc1e19f9bd4f/edit?invitationId=inv_7c1bdd58-2ffe-4d84-b6fa-b8738126262c&page=77UddvzHICQe#) for what we discussed
3. If you use the cursor `2024-09-29`, then you will find some patients like Anna (uid: `716e3fce-5483-4cce-99e9-cd2fae999a67`) where a different gender is specified in the encounter than on her patient record.

**Comment by mtuchi** [comment]
I need fresh eyes on this one, I think get-encounters step need to start before `upsert-teis` step i just need to figure out the mapping

**Comment by aleksa-krolls** [comment]
@mtuchi as discussed...
- let's add this step at the end so it can be turned on/off in future projects
- let's _update_ TEIs in bulk using the TEI id
- only update TEIs in this step _if_ the patient has an encounter where sex _is_ specified. (FYI I added another test patient `Yudi` that has 1 encounter where gender is _not specified_, so therefore he would be excluded from this update step)
![Image](https://github.com/user-attachments/assets/870afb0e-0c78-4258-a827-84dbdb3fc7db)

**Comment by mtuchi** [comment]
Hiya @aleksa-krolls i have added the `update-teis` step in the wf2. See #44. I have tested locally running only the `update-teis` step because `create-events` step is throwing some errors and i can't seem to figure out the issue. See error ⬇ 

```
 "conflicts": [
          {
            "object": "G0hLyxqgcO7",
            "value": "value_not_true_only"
          }
        ],
        "reference": "mPD397qms57"
```

**Comment by mtuchi** [comment]
Hiya @aleksa-krolls i have tested locally the whole workflow for `lastRunDateTime: "2024-09-30"` and it ran successfully and updated one trackedEntity. I have commented out `G0hLyxqgcO7` because it was causing an error when creating an event, please help troubleshoot that. Otherwise the PR #44 is ready review

**Comment by aleksa-krolls** [comment]
hey @mtuchi this is generally looking pretty good, but per the logs I'm wondering if we're mapping more encounters than we need to. Can you please check out my comments (yellow post-its) in [this "sex" diagram](https://lucid.app/lucidchart/38ae80f0-0650-4fea-9a11-dc1e19f9bd4f/edit?invitationId=inv_7c1bdd58-2ffe-4d84-b6fa-b8738126262c&page=77UddvzHICQe#) and lmk if we need to make any other changes?

Also fyi there is a known issue with some of the OMRS `sex` options missing from the optionMap... I am working with the client to resolve this.

**Comment by mtuchi** [comment]
@aleksa-krolls i have found the issue and worked on a fix, See my PR #46. You can merge and test again
